firedrake/supermeshimpl.pyx

```diff
@@ -64,6 +70,17 @@ def assemble_mixed_mass_matrix(V_A, V_B, candidates,
     vertex_map_A = mesh_A.coordinates.cell_node_map().values_with_halo
     vertex_map_B = mesh_B.coordinates.cell_node_map().values_with_halo
 
+    if mesh_A.comm.size > 1:
+        compatible = False
+        assert mesh_B._parallel_compatible is not None, "Whoever made mesh_B should explicitly mark mesh_A as having a compatible parallel layout."
+        for _mesh_A in mesh_B._parallel_compatible:
+            if mesh_A is _mesh_A():
+                compatible = True
+                break
+
+        if not compatible:
+            assert ValueError("Whoever made mesh_B should explicitly mark mesh_A as having a compatible parallel layout.")
```

Comment:
Does adding the check here break projection in the mesh hierarchy case? (we haven't said that the meshes are parallel compatible?)

Reply:
Since this function is an internal implementation one, we could move the checking into the interface code. That is in `intersection_finder` (which you already have), or else in the `supermesh.assemble_mixed_mass_matrix`?

Reply:
Yes, it does break the nested case.

I suppose by creating a `MeshHierarchy` one is implicitly asserting that the parallel layouts are compatible. So I'll leave these checks only in `intersection_finder`. I prefer to leave them there because someone might call `intersection_finder` without wanting to assemble the mixed mass matrix.

Reply:
Should I make a MeshHierarchy set `_parallel_compatible`? Or is that for meshing tools, like pragmatic?

Reply:
We decided to not do this last for now.